webclient/grpc/src/main/java/io/helidon/webclient/grpc/GrpcClientCall.java

```diff
@@ -0,0 +1,191 @@
+
+    private volatile Future<?> readStreamFuture;
+    private volatile Future<?> writeStreamFuture;
+
+    GrpcClientCall(GrpcClientImpl grpcClient, MethodDescriptor<ReqT, ResT> methodDescriptor, CallOptions callOptions) {
+        super(grpcClient, methodDescriptor, callOptions);
+        this.executor = grpcClient.webClient().executor();
+    }
+
+    @Override
+    public void request(int numMessages) {
+        LOGGER.finest(() -> "request called " + numMessages);
+        messageRequest.addAndGet(numMessages);
+        startReadBarrier.countDown();
+    }
+
+    @Override
+    public void cancel(String message, Throwable cause) {
+        LOGGER.finest(() -> "cancel called " + message);
+        responseListener().onClose(Status.CANCELLED, EMPTY_METADATA);
+        readStreamFuture.cancel(true);
+        writeStreamFuture.cancel(true);
+        close();
+    }
+
+    @Override
+    public void halfClose() {
+        LOGGER.finest("halfClose called");
+        sendingQueue.add(EMPTY_BUFFER_DATA);       // end marker
+    }
+
+    @Override
+    public void sendMessage(ReqT message) {
+        LOGGER.finest("sendMessage called");
+        BufferData messageData = BufferData.growing(BUFFER_SIZE_BYTES);
+        messageData.readFrom(requestMarshaller().stream(message));
+        BufferData headerData = BufferData.create(5);
+        headerData.writeInt8(0);                                // no compression
+        headerData.writeUnsignedInt32(messageData.available());         // length prefixed
+        sendingQueue.add(BufferData.create(headerData, messageData));
+        startWriteBarrier.countDown();
+    }
+
+    protected void startStreamingThreads() {
+        // write streaming thread
+        writeStreamFuture = executor.submit(() -> {
+            try {
+                startWriteBarrier.await();
+                LOGGER.fine("[Writing thread] started");
+
+                boolean endOfStream = false;
+                while (isRemoteOpen()) {
+                    LOGGER.finest("[Writing thread] polling sending queue");
+                    BufferData bufferData = sendingQueue.poll(WAIT_TIME_MILLIS, TimeUnit.MILLISECONDS);
+                    if (bufferData != null) {
+                        if (bufferData == EMPTY_BUFFER_DATA) {     // end marker
+                            LOGGER.finest("[Writing thread] sending queue end marker found");
+                            if (!endOfStream) {
+                                LOGGER.finest("[Writing thread] sending empty buffer to end stream");
+                                clientStream().writeData(EMPTY_BUFFER_DATA, true);
+                            }
+                            break;
+                        }
+                        endOfStream = (sendingQueue.peek() == EMPTY_BUFFER_DATA);
+                        boolean lastEndOfStream = endOfStream;
+                        LOGGER.finest(() -> "[Writing thread] writing bufferData " + lastEndOfStream);
```

Comment:
If there is concatenation, it should be guarded by an if, or use template (supported by socket context)

Reply:
Even inside a lambda?